django/db/backends/sqlite3/creation.py

```diff
@@ -106,3 +122,29 @@ def test_db_signature(self):
         else:
             sig.append(test_database_name)
         return tuple(sig)
+
+    def setup_worker_connection(self, _worker_id):
+        alias = self.connection.alias
+        worker_db = f"file:memorydb_{alias}_{_worker_id}?mode=memory&cache=shared"
+
+        start_method = multiprocessing.get_start_method()
+        if start_method == "fork":
+            source_db = sqlite3.connect(
+                f"file:memorydb_{alias}?mode=memory&cache=shared", uri=True
+            )
+        elif start_method == "spawn":
+            source_db = sqlite3.connect(
+                "file:%s_%s.sqlite3" % (alias, _worker_id), uri=True
+            )
+        else:
+            raise NotImplementedError(
+                f"Start method {start_method!r} is not supported."
+            )
+
+        second_db = sqlite3.connect(worker_db, uri=True)
+        source_db.backup(second_db)
+        source_db.close()
+        self.connection.settings_dict["NAME"] = worker_db
+        self.connection.connect()
+        second_db.close()
```

Comment:
So this section could do with a bit of thought. Looking at the Python sqlite [docs](https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.backup), it seems we should do:

```suggestion
        second_db = sqlite3.connect(worker_db, uri=True)
        with second_db:
            source_db.backup(second_db)
        source_db.close()

        # Re-open connection to in-memory database before closing copy connection.
        self.connection.settings_dict["NAME"] = worker_db
        self.connection.connect()
        second_db.close()
```

Not really sure what the context manager is for 🤷🏻‍♂️ 

Can we also rename `second_db` to `target_db`? While we're at it, perhaps we should also change the name of `worker_db` as it is just the connection string.

Other thoughts related to `.backup()` that might be worth exploring:

- It takes a `sleep` argument with a default of `0.25`. Maybe we can reduce this to speed up setup time?
- It takes a `progress` argument accepting a callable. Perhaps we can pass `verbosity` to `setup_worker_connection()` and display progress information?

Reply:
I'm not sure what the `with` block is for either. I read the PR earlier and that didn't make it any clearer for me as the implementation is written in c. 

The second example shows it written more like it is here, but I agree that it is a little untidy. 

I've not yet tested Nick's proposal, which I'd judge as more readable than the current attempt.

Reply:
I testted Nick's proposals I get a handful of test failures. I'm not entirely sure why. I therefore just progressed with the renaming suggestions. 

This still leaves the question about changing the other arguments to `backup()`

Reply:
FWIW, the `with ...` usage is documented as ["Connection objects can be used as context managers that automatically commit or rollback transactions. In the event of an exception, the transaction is rolled back; otherwise, the transaction is committed"](https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager), which is presumably useful and necessary when the data pages to be backed up are locked for writing etc (hence also, the sleep ...)?

Reply:
I looked at this again. When using the context manager I again saw test failures (different to those before) and they were intermittent. I hadn't seen that with the current patch. 

However given they are intermittent I'm not sure if it's an issue with the context manager or this patch.